### PR TITLE
[Feature] Content elements icons export

### PR DESCRIPTION
**Pull request by manuelbachl**
Added option in extension configuration to export set fontawesome icons aswell. If set, ext_localconf.php will be extended with the IconRegistry and the desired icons.

### CODE REVIEW

File: Classes/Aggregate/NewContentElementWizardAggregate.php [hunk 1]
```diff
@@ -52,6 +52,16 @@ class NewContentElementWizardAggregate extends AbstractAggregate implements Lang
...
     /**
+     * @var array
+     */
+    protected $icons = [];
+
+    /**
+     * @var array
+     */
+    protected $iconNames = [];
+
```
Comment (IchHabRecht) [inline Classes/Aggregate/NewContentElementWizardAggregate.php]: There is no need to add two extra properties here. Just take one associative array.

File: Classes/Aggregate/NewContentElementWizardAggregate.php [hunk 2]
```diff
@@ -92,7 +102,32 @@ protected function process()
 );
 
 EOS
+
         );
+
+        $extensionConfiguration = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['mask_export']);
```
Comment (IchHabRecht) [inline Classes/Aggregate/NewContentElementWizardAggregate.php]: There shouldn't be any need to unserialize the extension configuration multiple times. If the icon property is empty, no icon was registered.

File: Classes/Aggregate/NewContentElementWizardAggregate.php [hunk 3]
```diff
@@ -92,7 +102,32 @@ protected function process()
...
+        if (!empty($extensionConfiguration['exportIcons'])) {
+            $this->appendPhpFile(
+                'ext_localconf.php',
+<<<EOS
+
```
Comment (IchHabRecht) [inline Classes/Aggregate/NewContentElementWizardAggregate.php]: please remove this empty line

File: Classes/Aggregate/NewContentElementWizardAggregate.php [hunk 4]
```diff
@@ -92,7 +102,32 @@ protected function process()
...
+EOS
+            );
+
+            sort($this->icons);
+            foreach ($this->icons as $icon) {
+                $this->appendPhpFile(
+                    'ext_localconf.php',
+<<<EOS
+
```
Comment (IchHabRecht) [inline Classes/Aggregate/NewContentElementWizardAggregate.php]: please remove this extra line + see comments below

File: Classes/Aggregate/NewContentElementWizardAggregate.php [hunk 5]
```diff
@@ -113,11 +148,18 @@ protected function processElement(array $element)
             (!empty($element['description'])) ? $element['description'] : ''
         );
 
+        $extensionConfiguration = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['mask_export']);
+        if (!empty($extensionConfiguration['exportIcons'])) {
+            $iconName = substr($element['icon'], 3);
+        } else {
+            $iconName = 'content-textpic';
```
Comment (IchHabRecht) [inline Classes/Aggregate/NewContentElementWizardAggregate.php]: leave iconName empty if feature is not enabled. then set $this->icons[$iconName] = true;

Comment (IchHabRecht) [inline Classes/Aggregate/NewContentElementWizardAggregate.php]: Introduce and use $iconIdentifier = $iconName !== '' ? $iconName : 'content-textpic';

File: Classes/Aggregate/NewContentElementWizardAggregate.php [hunk 6]
```diff
@@ -113,11 +148,18 @@ protected function processElement(array $element)
...
+        }
+
         $this->appendPlainTextFile(
             $this->pageTSConfigFilePath . $this->pageTSConfigFileIdentifier,
 <<<EOS
             {$key} {
-                iconIdentifier = content-textpic
+                iconIdentifier = $iconName
```
Comment (IchHabRecht) [inline Classes/Aggregate/NewContentElementWizardAggregate.php]: use $iconIdentifier

File: Classes/Aggregate/NewContentElementWizardAggregate.php [hunk 7]
```diff
@@ -127,5 +169,23 @@ protected function processElement(array $element)
...
+                $contentElementIcon =
+<<<ICON
+\$iconRegistry->registerIcon(
+    '$iconName',
+    \TYPO3\CMS\Core\Imaging\IconProvider\FontawesomeIconProvider::class,
+    [
+        'name'     => '$iconName'
+    ]
+);
+ICON;
```
Comment (IchHabRecht) [inline Classes/Aggregate/NewContentElementWizardAggregate.php]: This code should be added on modifying the ext_localconf.php - Please only store iconName in the property. Not need to check for an existing one as same name should mean same icon

File: Classes/Aggregate/NewContentElementWizardAggregate.php [hunk 8]
```diff
@@ -127,5 +169,23 @@ protected function processElement(array $element)
 
 EOS
         );
+
+        if(!in_array($iconName, $this->iconNames)) {
+            $extensionConfiguration = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['mask_export']);
```
Comment (IchHabRecht) [inline Classes/Aggregate/NewContentElementWizardAggregate.php]: Please unserialize the extension configuration only once (see above).

File: Resources/Private/Language/locallang.xlf [hunk 9]
```diff
@@ -6,6 +6,9 @@
 			<trans-unit id="config.backendPreview">
 				<source>Add backend preview templates: If enabled the export contains some PHP and Fluid files to show a record preview in the page layout.</source>
 			</trans-unit>
+			<trans-unit id="config.exportIcons">
+				<source>Export content element icons: If enabled the export contains some PHP to export defined content element icons.</source>
```
Comment (IchHabRecht) [inline Resources/Private/Language/locallang.xlf]: You only added the feature to export defined fontawsome icons. By the way, is there any check which checks weather the icon was set in mask configuration? Please don't tell that you are exporting content element icons but the fontawsome icons

File: Resources/Private/Language/locallang.xlf [hunk 10]
```diff
@@ -6,6 +6,9 @@
 			<trans-unit id="config.backendPreview">
 				<source>Add backend preview templates: If enabled the export contains some PHP and Fluid files to show a record preview in the page layout.</source>
 			</trans-unit>
+			<trans-unit id="config.exportIcons">
```
Comment (IchHabRecht) [inline Resources/Private/Language/locallang.xlf]: Please use config.contentFontAwsomeIcons

File: ext_conf_template.txt [hunk 11]
```diff
@@ -1,2 +1,5 @@
 # cat=Experimental; type=boolean; label=LLL:EXT:mask_export/Resources/Private/Language/locallang.xlf:config.backendPreview
 backendPreview =
+
+# cat=Experimental; type=boolean; label=LLL:EXT:mask_export/Resources/Private/Language/locallang.xlf:config.exportIcons
+exportIcons =
```
Comment (IchHabRecht) [inline ext_conf_template.txt]: please use contentFontAwsomeIcons

**Comment by IchHabRecht** [comment]
Please also add an issue which your pull request belongs to. Thank you very much.

**Comment by IchHabRecht** [comment]
Hi Manuel,

would you mind to test #48 - this also cares about icons in preview folder.

**Comment by manuelbachl** [comment]
further development in #48

**Pull request closed without merging**